canvas_modules/common-canvas/src/palette/palette-flyout-utils.js

```diff
@@ -37,150 +37,173 @@ function getFilteredNodeTypeInfosByCategory(category, filterStrings) {
+// Ranking is based on:
+// * The number of hits across the 3 areas (category label, node label, description)
+// * Whether or not multiple filter strings (if more than one is provided)
+//   appear in any of the areas.
+function calcRanking(catLabelHitCounts, nodeLabelHitCounts, descHitCounts, filterStringsLength) {
 	let ranking = 0;
+	let multiStringHits = 0;
 
-	const totalHitCount = [filterStringsLength];
 	for (let i = 0; i < filterStringsLength; i++) {
-		totalHitCount[i] = catLabelHitCounts[i] + labelHitCounts[i] + descHitCounts[i];
-	}
+		ranking += catLabelHitCounts[i] + (10 * nodeLabelHitCounts[i]) + descHitCounts[i]; // Give extra weight to node label hit
 
-	let multiStringHits = 0;
-	for (let i = 0; i < totalHitCount.length; i++) {
-		ranking += totalHitCount[i];
-		if (totalHitCount[i] > 0) {
+		if (catLabelHitCounts[i] + nodeLabelHitCounts[i] + descHitCounts[i] > 0) {
 			multiStringHits++;
 		}
 	}
 
 	return ranking + (multiStringHits * 50);
 }
 
-// Returns an object containing an array of occurences and hit counts for the
+// Returns an object containing an array of occurrences and hit counts for the
 // main string passed in after being searched by the strings in the
 // filterStrings array passed in.
-function wordOccurences(mainString, filterStrings) {
-	let occurences = [];
+function wordOccurrences(mainString, filterStrings) {
+	let occurrences = [];
 	const hitCounts = [];
 	filterStrings.forEach((s) => {
 		if (s) {
-			const newOccurences = wordOccurencesByString(mainString, s);
-			occurences = normalize(occurences, newOccurences);
-			hitCounts.push(newOccurences.length > 0 ? 1 : 0);
+			const newOccurrences = wordOccurrencesByString(mainString, s);
+			occurrences = joinOccurrences(occurrences, newOccurrences, mainString);
+			hitCounts.push(newOccurrences.length > 0 ? 1 : 0);
 		}
 	});
-	return { occurences, hitCounts };
+	return { occurrences, hitCounts };
 }
 
-// Returns an array of occurences where the occurences from the two arrays
+// Returns an array of occurrences where the occurrences from the two arrays
 // passed in are joined together where they overlap so that the returned
-// array has no overlapping occurences.
-function normalize(occurences, newOccurences) {
-	if (occurences.length === 0) {
-		return newOccurences;
+// array has no overlapping occurrences.
+function joinOccurrences(occurrences, newOccurrences) {
+	if (occurrences.length === 0) {
+		return newOccurrences;
 	}
-	if (newOccurences.length === 0) {
-		return occurences;
+	if (newOccurrences.length === 0) {
+		return occurrences;
 	}
 
-	const addOccurences = [];
-	newOccurences.forEach((newOcc) => {
+	const addOccurrences = [];
+	newOccurrences.forEach((newOcc) => {
 		let handled = false;
-		occurences.forEach((occ) => {
-			if (newOcc.start >= occ.start && newOcc.start <= occ.end) {
-				if (newOcc.end > occ.end) {
+		occurrences.forEach((occ) => {
+			//   Occurrence         <-->
```

Comment:
I dont understand these comments, can you elaborate? Or are the comments required here?

Reply:
The algorithm is joining two arrays . Each array contains elements which describe an 'occurrence' of one of the filter strings in the main string being analyzed that is, either the category label, the node label or the description. The algorithm loops through the arrays looking for overlaps between occurrences so it can consolidate them together. So if, for example, I had and occurrence of { start: 3, end: 5 } and another of { start: 4, end: 7 } they would get consolidated into a single occurrence of  { start: 3, end: 7 }.

There are a few different combinations of this type of overlap which the `if` statements within the loop try to handle. The comments try to show the different possibilities of overlap that the associated `if` statement is handling.

I noticed actually that in globally correcting the spelling of `occurrence` it nudged the lines slightly out of place. So I have corrected that and added an additional line to show that the arrows are describing string occurrences. I'd like to keep the comments here even if they only make sense to me :)